### DISCUSSION TITLE
Web app architecture (actors all the way down!)

### DISCUSSION THREAD

**Discussion by ghost** [statelyai/xstate#1807] [General]
Hi & happy new year!

I stumbled XState a year or so ago and learned a lot reading the docs and discussions, so I owe David and the community a belated thank you for all the neat patterns and insights they've shared - thank you!

What brought me back to XState again was not FSMs, actually. It was the actor model which piqued my interest with its "_everything's_ an actor!" approach to systems along with properties like location transparency and simple message-passing and processing semantics. Seemed like a natural choice given the half-baked RPC API I was already implementing on top of a growing handful of services.

Most XState examples I've discovered seem to use it as part of an existing React app while I'd prefer to build out the app logic first then add rendering as a side effect. Are there any examples of web apps using XState from the ground up in this way? I imagine it'd look something like an actor hierarchy with an App root/supervisor and several child actors managing specific responsibilities like routing and rendering.

Of course, sanity check: is this a sane idea? CRUD UIs all seem to boil down to FSMs, so why not build the app as an FSM from the ground up, or at least extend use of the actor model from back to front?

Thanks again!

**Comment by davidkpiano** [comment]
> Are there any examples of web apps using XState from the ground up in this way?

Sort of... there are examples of using XState without any framework, such as the [XState Frontend Masters workshop repo (free)](https://github.com/davidkpiano/frontend-masters-xstate-workshop), but the exercises don't do anything too complex with actors.

> Of course, sanity check: is this a sane idea?

It's a very sane idea, and it's actually how I'm building apps today - React is relegated to just a view layer. IMO, the XState TodoMVC example is actually a good example of this: https://codesandbox.io/s/xstate-todomvc-33wr94qv1

The majority of the logic lives in `todosMachine` and `todoMachine`, instances of which are spawned from `todosMachine`. So it's actually an actor system of todos that are rendered with React, but it wouldn't be much effort to use a different view library for rendering.

---

I'm also exploring building out the inspector/creator (future) to do exactly what you're describing -- model your app logic as an actor system, use the visualizer to "interact" with the app to verify the correct behavior, and only at the very end, add a view layer (such as React or Vue) to render the app.

It's sort of like the Matrix - you learn to "see" the boxes and arrows as the app before it's actually rendered. Your intuition is correct (IMO) that all apps can be represented this way. The view layer should only be a thin, easily interchangeable layer.

**Reply by ghost** [reply]
Thank you for the thoughtful response. My original post was a bit of a brain dump after a long drive, so I'd like to try to clarify what I had in mind.

The gist of it is wrapping up the whole client app in an XState machine definition, from initialization to module loading to error handling, such that the entry point is something like interpret(App), not SomeViewLib.render(<Component/>). Each app module would be some variant of an invoked service, e.g. a rendering service, assuming strictly functional, subscribes to parent state updates and re-renders on changes.

My goal is to collapse complexity by unifying the operating model. The combination of actor model and explicit statecharts seems to me like the ultimate UI red pill, trading a little bit of verbosity for consistency and intelligibility.

What I'm not clear on is if this is another shiny-object, silver-bullet moment for me or simply taking the principles behind XState to their logical end.

P.S. Another source of enthusiasm is the potential for modeling workflows/business processes as statecharts...

**Comment by redbar0n** [comment]
What about using XState on the backend, to make a loosely coupled Modular Monolith? The Actor Model should be well suited for that (and for later more easily transitioning to full on distributed microservices).

But what would that look like?

Any reference implementation or backend framework in Node/Deno that uses XState to make a Modular Monolith?

**Reply by davidkpiano** [reply]
I've started experimenting with this in the past using XActor: https://github.com/statelyai/xactor
